cmd/validate.go

```diff
@@ -0,0 +1,272 @@
+	// at each level once, and stop when we've fully explored the schema. If there are errors, print
+	// all the detail we have about them. Use an unbuffered channel to register errors as they come in.
+	erred := 0
+	errs := make(chan error)
+	go func() {
+		for err := range errs {
+			erred++
+			cmdutil.ErrPrintf("%v\n", err)
+		}
+	}()
+
+	// We use a worker pool to limit work-in-progress, and a wait group to know when all queued work
+	// is complete. Put the plugin on the worker pool. Listing the registry should never error.
+	wp := newPool(parallel)
+	entries, _ := plugin.CachedList(ctx, registry)
+	for _, e := range entries {
+		wp.Submit(func() { processEntry(ctx, pw, wp, e, all, errs) })
+	}
+
+	// Wait for the workgroup's we've queued to finish, then stop the worker pool.
+	wp.Finish()
+	// Leave time for progress to finish rendering.
+	time.Sleep(100 * time.Millisecond)
+	pw.Stop()
+	if erred > 0 {
+		cmdutil.ErrPrintf("Found %v errors.\n", erred)
+		return exitCode{1}
+	}
+
+	cmdutil.Println("Looks good!")
+	return exitCode{0}
+}
+
+type pool struct {
+	wp *workerpool.WorkerPool
+	wg *sync.WaitGroup
+}
+
+func newPool(parallel int) pool {
+	return pool{wp: workerpool.New(parallel), wg: &sync.WaitGroup{}}
+}
+
+func (p pool) Submit(f func()) {
+	p.wg.Add(1)
+	p.wp.Submit(f)
+}
+
+func (p pool) Done() {
+	p.wg.Done()
+}
+
+func (p pool) Finish() {
+	// Wait for the workgroup's we've queued to finish, then stop the worker pool.
+	p.wg.Wait()
+	p.wp.StopWait()
+}
+
+type criteria struct {
+	list, read, stream, exec bool
+}
+
+func newCriteria(entry plugin.Entry) criteria {
+	var crit criteria
+	crit.list = plugin.ListAction().IsSupportedOn(entry)
+	crit.read = plugin.ReadAction().IsSupportedOn(entry)
+	crit.stream = plugin.StreamAction().IsSupportedOn(entry)
+	crit.exec = plugin.ExecAction().IsSupportedOn(entry)
+	return crit
+}
+
+func (c criteria) String() string {
+	s := []byte("----")
+	if c.list {
+		s[0] = 'l'
+	}
+
+	if c.read {
+		s[1] = 'r'
+	}
+
+	if c.stream {
+		s[2] = 's'
+	}
+
+	if c.exec {
+		s[3] = 'x'
+	}
+	return string(s)
+}
+
+func processEntry(ctx context.Context, pw progress.Writer, wp pool, e plugin.Entry, all bool, errs chan<- error) {
+	defer wp.Done()
+	name := plugin.ID(e)
+	crit := newCriteria(e)
+	tracker := progress.Tracker{Message: fmt.Sprintf("Testing %s %s", crit, name), Total: 4}
+	pw.AppendTracker(&tracker)
+
+	if plugin.ListAction().IsSupportedOn(e) {
+		entries, err := plugin.CachedList(ctx, e.(plugin.Parent))
+		if err != nil {
+			errs <- formatErr(fmt.Sprintf("Error validating List on %v", name), err)
+			return
+		}
+
+		if all {
+			for _, entry := range entries {
+				// Make a local copy for the lambda to capture.
+				entry := entry
+				wp.Submit(func() { processEntry(ctx, pw, wp, entry, all, errs) })
+			}
+		} else {
+			// Group children by ones that look "similar", and select one from each group to test.
+			groups := make(map[criteria][]plugin.Entry)
```

Comment:
Do you think we should require external plugins to return a TypeID? That would eliminate the "group on criteria" stuff.

I was originally thinking a TypeID would be required if the external plugin supports entry schemas, but with `validate`, I'm starting to think we should just require it. It's also less things people need to juggle ("Oh I have to include a TypeID whenever I'm using schemas. I'm not using schema stuff so no need for me to do that." vs. "Oh, I gotta include a TypeID. Ok, let me make something up"). Although "make something up" could be an "easier said than done" thing. So dunno what's best here.

Reply:
TypeID would be useful, but not sure we should require it.

Reply:
Ok, that sounds good.